### PR TITLE
Gemma3 vision attention optimization

### PR DESCRIPTION
**Pull request by mstojkoTT**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/28517) - covers part of PR only

#### Optimisations
- using `ttnn.experimental.nlp_create_qkv_heads` to create QKV heads, which helps us avoid a bunch of transposes and reshape which were slow and also caused hangs. We have just 1 operation instead of a reshape, transpose and matmul, for each of the 3 QKV (so 1 op instead of 9)
- we now have 1 CCL less (2 allgathers removed, 1 allreduce instad of them, which is actuall allgather under the hood)
- It used to be that all broadcast in `models/demos/gemma3/tt/gemma_image_block.py` (`TtGemmaImageTransformerBlock`) after all gather was under the hood being done by all broadcast which is unoptimal. This is no longer the case (that all gather is gone entirely). We wanted to avoid this allbroadcast as it is considered slower than regular allgather
- removed unused `wq, wk, wv` in `models/demos/gemma3/tt/gemma_image_attention.py`, we are now using the fused weights. These weights were completely unused and were using, perhaps approximately 50% of the model. (techincally they were used but the fused ones existed and they were unused)
- changed `self.bo` (bias) in gemma attention class from sharded to replicated, as it's now being added after the CCL (allreduce). (this was required because of the different way the matmul works now)
    - `self.bo` now addeda after the matmul (if we added it before the matmul it would get multiplied by eg. 8 after the allreduce on T3K)
- same point as above goes for `bq, bk, bv`
- weights of for attention (Q K V weights) are now being sharded across the other dimension which requires an allreduce to sum up the results after matmul-ing

#### Tests
- `models/demos/gemma3/tests/test_vision_attention.py` - this test now expects replicated activations as attention class has changed (no longer outputs sharded activations)
- added weights cache path to tests `models/demos/gemma3/tests/test_vision_attention.py`, `models/demos/gemma3/tests/test_vision_cross_attention_transformer.py`

### Results

- perf for TtGemmaImageTransformerBlock` almost twice as fast (from 22.7 microsends to 12.5 microseconds)
- This unfortunately does not translate to the same perf increase for the entire vision model. The perf for the entire vision model went from `0.67` to `0.45` for a 896x896 RGB image (as definied in models/demos/gemma3/tests/test_vision_cross_attention_transformer.py)
- The memory for the attention module (`TtGemmaImageAttention) is around 2 times less (x~x/2). (not measured but seems obvious)

Old perf for gemma3 27b on T3K for `TtGemmaImageTransformerBlock` forward():
image

New perf for the same (gemma3 27b on T3K for `TtGemmaImageTransformerBlock` forward()):
image

### Checklist and tests
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes

https://github.com/tenstorrent/tt-metal/actions/runs/19297860191
https://github.com/tenstorrent/tt-metal/actions/runs/19297857636
https://github.com/tenstorrent/tt-metal/actions/runs/19297849793
https://github.com/tenstorrent/tt-metal/actions/runs/19297845016

### CODE REVIEW

File: models/demos/gemma3/tests/test_vision_attention.py [hunk 1]
```diff
@@ -71,13 +71,11 @@ def test_attention_inference(batch, num_chunks, mesh_device, reset_seeds):
 
     tt_out = tt_model(attention_input)
 
-    # Doing contract in tt is correct!!
-    tt_output_torch = ttnn.to_torch(
-        tt_out, mesh_composer=ttnn.ConcatMeshToTensor(mesh_device, dim=-1), device=mesh_device
-    )[0, :, :, :]
+    # Tensor is replicated, so just get from first device
+    tt_output_torch = ttnn.to_torch(ttnn.get_device_tensors(tt_out)[0])[0, :, :, :]
```
Comment (mstojkoTT) [inline models/demos/gemma3/tests/test_vision_attention.py:75]: Had to update this as now gemma attention outputs a replicated tensor

File: models/demos/gemma3/tests/test_vision_attention.py [hunk 2]
```diff
@@ -71,13 +71,11 @@ def test_attention_inference(batch, num_chunks, mesh_device, reset_seeds):
...
 
     reference_output = reference_model(pt_attention_input)[0]
-    tt_output_torch = tt_output_torch[:, :4097, :]
+    tt_output_torch = tt_output_torch[:, :seq_len, :]
```
Comment (mstojkoTT) [inline models/demos/gemma3/tests/test_vision_attention.py:78]: Small cleanup for the test, outside of these optimisations

File: models/demos/gemma3/tt/gemma_image_block.py [hunk 3]
```diff
@@ -102,23 +102,36 @@ def forward(self, x_11SH, mask=None):
...
+        # self.c_fc_bias = as_interleaved_tensor("c_fc", "bias", ttnn.bfloat16, dim=-1)
+        # self.c_fc_bias = ttnn.reshape(self.c_fc_bias, [1, -1])
+        # self.c_proj_weight = as_interleaved_tensor("c_proj", "weight", dtype, dim=-2)
+        # self.c_proj_bias = as_interleaved_tensor("c_proj", "bias", ttnn.bfloat16, dim=None)
+
         attn_out = self.attn(self.ln_1(x_11SH), mask=mask)
         if self.gated:
             attn_out = ttnn.mul(attn_out, ttnn.tanh(self.gate_attn))
 
-        if self.num_devices > 1:
```
Comment (mstojkoTT) [inline models/demos/gemma3/tt/gemma_image_block.py:109]: Attention now already returns tensor that is the same across all devices (after the all reduce) so that is why this is removed

File: models/demos/gemma3/tt/gemma_image_attention.py [hunk 4]
```diff
@@ -292,45 +227,32 @@ def forward(self, x_11SH, mask=None):
         seq_len = x_11SH.shape[-2]
         batch_size = x_11SH.shape[0]
 
+        if len(x_11SH.shape) == 3:
```
Comment (ppetrovicTT) [inline models/demos/gemma3/tt/gemma_image_attention.py:233]: why is this needed?

File: models/demos/gemma3/tt/gemma_image_attention.py [hunk 5]
```diff
@@ -90,6 +90,7 @@ def pad_head_dim(weight, heads_out=True):
         wk_padded = pad_head_dim(self.state_dict[wk_str])
         wv_padded = pad_head_dim(self.state_dict[wv_str])
         wo_padded = pad_head_dim(self.state_dict[wo_str], heads_out=False)
+
         wq_chunked, wk_chunked, wv_chunked = (
```
Comment (ppetrovicTT) [inline models/demos/gemma3/tt/gemma_image_attention.py:94]: i wonder if padding as above and chunking by this dimension is optimal. it seems to me that in `tt_transformers/tt/attention.py` it is being chunked by a different dimension. have you tried different configurations?

File: models/demos/gemma3/tt/gemma_image_attention.py [hunk 6]
```diff
@@ -366,24 +289,9 @@ def forward(self, x_11SH, mask=None):
...
-                multi_device_global_semaphore=self.tt_ccl.get_and_cycle_ag_semaphore_handles(),
-                num_links=1,
-                topology=ttnn.Topology.Ring,
-                barrier_semaphore=self.tt_ccl.get_and_cycle_barrier_semaphore_handle(),
-                chunks_per_sync=10,
-                num_workers_per_link=2,
-                num_buffers_per_channel=2,
-            )
-
         output_11SH = ttnn.linear(
```
Comment (ppetrovicTT) [inline models/demos/gemma3/tt/gemma_image_attention.py:296]: why is bias removed from this `ttnn.linear`?
i see you've added it later, but i don't understand the reason.

**Comment by ppetrovicTT** [comment]
@pavlepopovic @ipotkonjak-tt we would appreciate your input on proposed optimizations. Thanks!

**Review by ipotkonjak-tt — APPROVED** [review]
lgtm

File: models/demos/gemma3/tt/gemma_image_attention.py [hunk 7]
```diff
@@ -393,4 +301,38 @@ def forward(self, x_11SH, mask=None):
...
+        return output_after_bias
+
+
+def _all_reduce_helper(caller, input_tensor):
+    w2_out_gathered = ttnn.experimental.all_gather_async(
+        input_tensor,
+        persistent_output_buffer=None,
+        dim=1,
+        multi_device_global_semaphore=caller.tt_ccl.get_and_cycle_ag_semaphore_handles(),
+        num_links=1,
```
Comment (pavlepopovic) [inline models/demos/gemma3/tt/gemma_image_attention.py]: note - is there a way to auto deduce num_links? (e.g. on 6u it will be 4, n300/t3k it is 1). If we stamp this model 4 times on a 6u, we can use 4 links there and get better perf?

Comment (pavlepopovic) [reply models/demos/gemma3/tt/gemma_image_attention.py]: Also - I see AG is still the most of model cycles - is it using bfloat16 or bfp8 inputs/outputs? If we're using bfloat16 then we can potentially reduce the data transfered by ~2x by switching to bfp8, but don't know how will that reflect on accuracy

**Review by pavlepopovic — APPROVED** [review]
nice :)

File: models/demos/gemma3/tt/gemma_image_attention.py (hunk 5 above)
Comment (mstojkoTT) [reply models/demos/gemma3/tt/gemma_image_attention.py:94]: I have tried different paddings and I also tried no padding but it did not work, as I mentioned offline. SDPA and create_heads operator were giving me errors. It is definetly worth trying more. Will look at this file you suggested for further optimisations (not for this PR though). This is definetly another low hanging fruit to be optimised.

File: models/demos/gemma3/tt/gemma_image_attention.py (hunk 4 above)
Comment (mstojkoTT) [reply models/demos/gemma3/tt/gemma_image_attention.py:233]: 2 reasons:
- Because the rest of the model does not give it this additional shape (it gives for example (1, 4096, 1152) instead of (1, 1, 4096, 1152)) and some of the operators later on expect this shape. If I remember correctly `ttnn.experimental.nlp_create_qkv_heads` a few lines below throws errors if this is not the case
- some of the tests (`models/demos/gemma3/tests/test_vision_attention.py` and `models/demos/gemma3/tests/test_vision_transformer_block.py`) give it this "correct" shape (eg. (1, 1, 4096, 1152)), while when running the model end to end it gives it the "wrong" shape.

I choose to solve it this way (this unsqueeze line). It's pretty ugly. Any suggestions for a better way?

File: models/demos/gemma3/tt/gemma_image_attention.py (hunk 6 above)
Comment (mstojkoTT) [reply models/demos/gemma3/tt/gemma_image_attention.py:296]: Because this matmul is split among different devices and later on the results are added (via `_all_reduce_helper` line below), if I added the bias here it would be added 8 times in the case of the T3K after the all reduce.

I had an alternative option:
- bias = bias / num_devices. This felt a bit dirty since I'm changing the weights/biases of the model at runtime 

I chose to do this instead because this bias division felt harder to debug later perhaps (as the bias/weight changes at runtime depending on the device)

File: models/demos/gemma3/tt/gemma_image_attention.py (hunk 7 above)
Comment (mstojkoTT) [reply models/demos/gemma3/tt/gemma_image_attention.py]: Thanks for the feedback. Will definetly try that out for the next optimisations.

Comment (mstojkoTT) [reply models/demos/gemma3/tt/gemma_image_attention.py]: The num_links thing I will implement now

File: models/demos/gemma3/tt/gemma_image_attention.py (hunk 4 above)
Comment (ppetrovicTT) [reply models/demos/gemma3/tt/gemma_image_attention.py:233]: You're converting [Batch, Seqlen, Dim] into [1, B, S, D] instead of [B, 1, S, D].
Check out nlp_create_qkv_heads pybind description.

There's already `batch_size` and `seq_len`, so you can just:
```
ttnn.reshape(x_11SH, (batch_size, 1, seq_len, -1))
```
(-1 will infer the dimension)

Comment (ppetrovicTT) [reply models/demos/gemma3/tt/gemma_image_attention.py:233]: also, please put a comment why is that needed:
```
ttnn.embedding will return [b, s, d] and nlp_create_qkv_heads expects [b, 1, s, d]
```

File: models/demos/gemma3/tt/gemma_image_attention.py [hunk 8]
```diff
@@ -393,4 +301,38 @@ def forward(self, x_11SH, mask=None):
...
+            output_all_reduce = _all_reduce_helper(self, output_11SH)
+            ttnn.deallocate(output_11SH)
+        else:
+            output_all_reduce = output_11SH
+
+        if self.bo is not None:
+            output_after_bias = ttnn.add(output_all_reduce, self.bo)
+        else:
+            output_after_bias = output_all_reduce
+        ttnn.deallocate(output_all_reduce)
```
Comment (ppetrovicTT) [inline models/demos/gemma3/tt/gemma_image_attention.py]: have you tried the code path without `self.bo`?
i'm worried that deallocate will actually clean up output_after_bias as well.

i've already had some issues that the assignment is in python only and that ttnn reference were not calculated, so deallocate will remove the tensor even if we think we're referencing it. not sure if this is fixed in ttnn.

Comment (ppetrovicTT) [reply models/demos/gemma3/tt/gemma_image_attention.py]: and i find it interesting that this concern is actually addressed above for deallocate(output_11SH). how come these are different?

**Review by ppetrovicTT — COMMENTED** [review]
You probably don't want to update LLM and UMD.

File: models/demos/gemma3/tt/gemma_image_attention.py (hunk 6 above)
Comment (ppetrovicTT) [reply models/demos/gemma3/tt/gemma_image_attention.py:296]: there's a new commit from https://github.com/tenstorrent/tt-metal/pull/31649. please check how it impacts your change.

**Review by nostojicTT — APPROVED** [review]

File: models/demos/gemma3/tt/gemma_image_attention.py (hunk 6 above)
Comment (mstojkoTT) [reply models/demos/gemma3/tt/gemma_image_attention.py:296]: It appears to me that it does not impact it. Using this bias in ttnn::linear would perhaps result in a tiny bit faster inference but given that our op to op gaps are not big I think it is not that important that we use this fused variant.

File: models/demos/gemma3/tt/gemma_image_attention.py (hunk 8 above)
Comment (mstojkoTT) [reply models/demos/gemma3/tt/gemma_image_attention.py]: I fixed  the code. 

`how come these are different?` - just my mistake, didn't check it

File: models/demos/gemma3/tt/gemma_image_attention.py [hunk 9]
```diff
@@ -292,45 +227,34 @@ def forward(self, x_11SH, mask=None):
         seq_len = x_11SH.shape[-2]
         batch_size = x_11SH.shape[0]
 
+        # Reason this line is needed - ttnn.embedding will return [b, s, d] and nlp_create_qkv_heads expects [b, 1, s, d]
```
Comment (ppetrovicTT) [inline models/demos/gemma3/tt/gemma_image_attention.py]: Nitpick:
> Reason this line is needed

Hahah, this sentence itself is not needed 😁 

I would put:

> Reshape required:
> ttnn.embedding returns [b, s, d]
> ttnn.nlp_create_qkv_heads expects [b, 1, s, d]

**Review by ppetrovicTT — APPROVED** [review]
